### PR TITLE
feat(fetch): expose timings

### LINKED ISSUE #19621: [Question] Is there a way to return an api request response time?
I am using the request fixture for API testing and I am just wondering is there a way to find what the api response time was?

### PR DESCRIPTION
**Pull request by Skn0tt**
Closes https://github.com/microsoft/playwright/issues/19621. Adds the same `timings()` we have for browser responses to `APIResponse`. Please apply some extra care in reviewing the timing calculations.

In the protocol change, I was unsure wether to extend the existing `ResourceTiming` type or to add another property. I went with the added property to be in-line with how the events work for browser requests - let me know if we should do it differently instead.

### CODE REVIEW

File: docs/src/api/class-request.md [hunk 1]
```diff
@@ -314,7 +314,7 @@ Returns resource size information for given request.
   - `startTime` <[float]> Request start time in milliseconds elapsed since January 1, 1970 00:00:00 UTC
   - `domainLookupStart` <[float]> Time immediately before the browser starts the domain name lookup for the
     resource. The value is given in milliseconds relative to `startTime`, -1 if not available.
-  - `domainLookupEnd` <[float]> Time immediately after the browser starts the domain name lookup for the resource.
+  - `domainLookupEnd` <[float]> Time immediately after the browser ends the domain name lookup for the resource.
```
Comment (Skn0tt) [inline docs/src/api/class-request.md:317]: this is a drive-by fix - pretty sure it shouldn't say "start the domain name lookup"

File: packages/playwright-core/src/server/fetch.ts [hunk 2]
```diff
@@ -422,14 +422,35 @@ export abstract class APIRequestContext extends SdkObject {
...
         const notifyBodyFinished = () => {
+          function relativeTime(time: number | undefined): number {
+            if (!time)
+              return -1;
+            return time - startAt;
+          }
+          const endAt = monotonicTime();
+          // spec: https://developer.mozilla.org/en-US/docs/Web/API/PerformanceResourceTiming
+          const timing: channels.ResourceTiming = {
+            startTime: startAt,
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:433]: The docs say `startTime` is a wall time, not monotonic time.

File: packages/playwright-core/src/server/fetch.ts [hunk 3]
```diff
@@ -422,14 +422,35 @@ export abstract class APIRequestContext extends SdkObject {
...
+            domainLookupStart: dnsLookupAt ? 0 : -1,
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:434]: Why zero and not `relativeTime()`?

File: packages/playwright-core/src/server/fetch.ts [hunk 4]
```diff
@@ -422,14 +422,35 @@ export abstract class APIRequestContext extends SdkObject {
...
+            domainLookupEnd: dnsLookupAt ? dnsLookupAt! - startAt : -1,
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:435]: `relativeTime()`?

File: packages/playwright-core/src/server/fetch.ts [hunk 5]
```diff
@@ -422,14 +422,35 @@ export abstract class APIRequestContext extends SdkObject {
...
           notifyRequestFinished(body);
           fulfill({
             url: response.url || url.toString(),
             status: response.statusCode || 0,
             statusText: response.statusMessage || '',
             headers: toHeadersArray(response.rawHeaders),
-            body
+            body,
+            timing,
+            responseEndTiming,
```
Comment (dgozman) [inline packages/playwright-core/src/server/fetch.ts:453]: Why do we have two sets of timings now?

File: packages/playwright-core/src/server/fetch.ts (hunk 3 above)
Comment (Skn0tt) [reply packages/playwright-core/src/server/fetch.ts:446]: My understand is that the browser has some other steps like cache resolution before the DNS lookup, so there might be time between startTime and the DNS lookup start. On Node.js, I don't think there's anything between that - so it's zero, because we know the DNS lookup happens immediately after the request start.

File: packages/playwright-core/src/server/fetch.ts (hunk 5 above)
Comment (Skn0tt) [reply packages/playwright-core/src/server/fetch.ts:465]: For browser requests, most of the timings are transmitted in the `Response` type, and then `responseEndTiming` arrives later in the `requestFinished` and `requestFailed` events. I opted to make this similar, so we have a big set of timings in `timing` and the final timing in `responseEndTiming`.

I thought about amending the `ResourceTiming` type instead, but then `requestFinished` would suddenly have the response end time both in `responseEndTiming` and in `response.timing.responseEnd` - that felt confusing.

File: packages/playwright-core/src/server/fetch.ts (hunk 4 above)
Comment (Skn0tt) [reply packages/playwright-core/src/server/fetch.ts]: done in https://github.com/microsoft/playwright/pull/32647/commits/b1d523b009dc99d856a9dd870dfd639f4671fd00

File: packages/playwright-core/src/server/fetch.ts (hunk 2 above)
Comment (Skn0tt) [reply packages/playwright-core/src/server/fetch.ts]: done in https://github.com/microsoft/playwright/pull/32647/commits/b717257013f459955f0172483e1b52b55ac7305d

**Comment by Skn0tt** [comment]
We discussed this with the team and decided against exposing this via the API. Playwright isn't a network performance testing tool, and we don't want people using it like one. Rough timings can easily be measured in userland.

I'll open a separate PR to fix the bugs around HTTP / SOCKS Proxy we found in the existing measurements for HAR timings.
